plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java

```diff
@@ -125,12 +128,32 @@ public static Optional<String> getFilterClause(TupleDomain<ColumnHandle> tupleDo
         return Optional.empty();
     }
 
-    private static String toPredicate(PinotColumnHandle pinotColumnHandle, Domain domain)
+    private static Optional<String> toPredicate(PinotColumnHandle pinotColumnHandle, Domain domain)
     {
         String predicateArgument = pinotColumnHandle.isAggregate() ? pinotColumnHandle.getExpression() : quoteIdentifier(pinotColumnHandle.getColumnName());
+        ValueSet valueSet = domain.getValues();
+        if (valueSet.isNone()) {
+            verify(!domain.isNullAllowed(), "IS NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
+            return Optional.of(format("(%s != %s)", predicateArgument, predicateArgument));
+        }
+        if (valueSet.isAll()) {
+            verify(domain.isNullAllowed(), "IS NOT NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
+            // Pinot does not support "1 = 1" syntax: see https://github.com/apache/pinot/issues/10600
+            // As a workaround, skip adding always true to conjuncts
+            return Optional.empty();
+        }
+        verify(!domain.getValues().getRanges().getOrderedRanges().isEmpty() && !domain.isNullAllowed(), "IS NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
         List<String> disjuncts = new ArrayList<>();
         List<Object> singleValues = new ArrayList<>();
-        for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+        boolean invertPredicate = false;
+        if (!valueSet.isDiscreteSet()) {
+            ValueSet complement = domain.getValues().complement();
+            if (complement.isDiscreteSet()) {
+                invertPredicate = complement.isDiscreteSet();
+                valueSet = complement;
+            }
+        }
+        for (Range range : valueSet.getRanges().getOrderedRanges()) {
```

Comment:
For my understanding, these changes are for null handling or for additional predicate types ?

Reply:
This is for `!=` and `NOT IN` - I noticed the issue when adding back the test.